cvat/apps/engine/rq.py

```diff
@@ -0,0 +1,527 @@
+    ORG_ID = "org_id"
+    ORG_SLUG = "org_slug"
+    STATUS = "status"
+    PROGRESS = "progress"
+    TASK_PROGRESS = "task_progress"
+    # export specific fields
+    RESULT_URL = "result_url"
+    RESULT = "result"
+    FUNCTION_ID = "function_id"
+    EXCEPTION_TYPE = "exc_type"
+    EXCEPTION_ARGS = "exc_args"
+    TMP_FILE = "tmp_file"
+
+
+class WithMeta(Protocol):
+    meta: dict[str, Any]
+
+
+class _AbstractRQMetaAttribute:
+    def __init__(
+        self, key: str, *, optional: bool = False, validator: Callable | None = None
+    ) -> None:
+        assert validator is None or callable(validator), "validator must be callable"
+        self._key = key
+        self._validator = validator
+        self._optional = optional
+
+
+class _GettableRQMetaAttribute(_AbstractRQMetaAttribute):
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        if self._optional:
+            return instance.meta.get(self._key)
+
+        return instance.meta[self._key]
+
+
+class _SettableRQMetaAttribute(_AbstractRQMetaAttribute):
+    def validate(self, value):
+        if value is None and not self._optional:
+            raise ValueError(f"{self._key} is required")
+        if value is not None and self._validator and not self._validator(value):
+            raise ValueError("Wrong type")
+
+    def __set__(self, instance: WithMeta, value: Any):
+        self.validate(value)
+        instance.meta[self._key] = value
+
+
+class ImmutableRQMetaAttribute(_GettableRQMetaAttribute):
+    pass
+
+
+class MutableRQMetaAttribute(_GettableRQMetaAttribute, _SettableRQMetaAttribute):
+    pass
+
+
+class UserRQMetaAttribute(ImmutableRQMetaAttribute):
+    def __init__(self, *, optional: bool = False, validator: Callable | None = None) -> None:
+        super().__init__(RQJobMetaField.USER, optional=optional, validator=validator)
+
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        assert RQJobMetaField.USER == self._key
+        return UserMeta(instance.meta[self._key])
+
+
+class RequestRQMetaAttribute(ImmutableRQMetaAttribute):
+    def __init__(self, *, optional: bool = False, validator: Callable | None = None) -> None:
+        super().__init__(RQJobMetaField.REQUEST, optional=optional, validator=validator)
+
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        assert RQJobMetaField.REQUEST == self._key
+        return RequestMeta(instance.meta[self._key])
+
+
+class UserMeta:
+    id: int = ImmutableRQMetaAttribute(
+        RQJobMetaField.UserField.ID, validator=lambda x: isinstance(x, int)
+    )
+    username: str = ImmutableRQMetaAttribute(
+        RQJobMetaField.UserField.USERNAME, validator=lambda x: isinstance(x, str)
+    )
+    email: str = ImmutableRQMetaAttribute(
+        RQJobMetaField.UserField.EMAIL, validator=lambda x: isinstance(x, str)
+    )
+
+    def __init__(self, meta: dict[RQJobMetaField.UserField, Any]) -> None:
```

Comment:
This type annotation is incorrect, because e.g. `UserField.ID` is not an instance of `UserField`.